docs/architecture/adr-001-initial-QGB-p2p-network.md

```diff
@@ -0,0 +1,195 @@
+- Put the Signature index in the DHT.
+
+```go
+attestationConfirm := AttestationConfirm {
+...
+}
+cid, err := cidlink.DefaultLinkSystem().Store(..., attestationConfirm)
+signatureIndex := SignatureIndex {
+  CID: cid,
+  // other fields
+}
+indexKey := "<nonce,orchestrator_addr,type>"
+dht.PutValue(ctx, indexKey, signatureIndex)
+```
+
+The DHT validator should run `ValidateBasics` on the `SignatureIndex` because if a value with a certain key was posted, it can't be overwritten.
+
+**Consumers**: Relayer, which will be querying attestations:
+
+- Create the `SignatureIndex`. All the fields would be known when querying for a certain signature.
+- Query the signature index from the DHT
+- Query the confirm from bitswap
+
+```go
+indexKey := "<nonce,orchestrator_addr,type>"
+index, err := dht.GetValue(ctx, indexKey)
+confirm, err := exchange.GetBlock(ctx, index.CID)
+```
+
+### Limitations
+
+Using the above approach will give us the following limitations:
+
+- Consumers will have to know when to query the network:
+  - This won't be an issue since we can define a delay to start querying: the time when the attestation request was created in the state machine + a few minutes
+- We will need to handle republishing the signature indexes periodically. This is because the DHT has a delay of 48 hours before it deletes the values put on it.
+
+### Options 2: DHT alone
+
+Instead of using bitswap along with the DHT, we could use only the DHT to keep the signatures as values.
+
+This will have the same limitations as option 1. However, these shouldn't be an issue with our use case.
+
+Also, it won't be much of a difference between using the DHT as an indexing mechanism, in terms of storage as the structs are almost similar.
+
+### Option 3: [Conflict-Free Replicated Data Types](https://github.com/ipfs/go-ds-crdt)
+
+- Merkle CRDT uses the following components:
+  - A DAG-syncer: which syncs DAG nodes between nodes and publishes new ones. IPFS will be used for this.
+  - A broadcaster: which sends messages to all the peers. This will be LibP2P PubSub.
+- Limitations:
+  - Ever-growing DAG: shouldn't be a problem in our case since we only have small unrelated key values.
+  - Merkle clock sorting: having access to a logical clock to sort events as they happen. This shouldn't cause us any issue since the signatures are not order dependent.
+  - TBD.
+- We should take a deeper look at the repo [go-ds-crdt](https://github.com/ipfs/go-ds-crdt) as there isn't much activity:
+  - low number of issues.
+  - has a single contributor doing most of the work.
+  - No production-grade project is using it.
+  - depends on [ipfs-lite](https://github.com/hsanjuan/ipfs-lite) which is maintained by the same person. The alternative would be to use the official IPFS daemon.
+
+### Option 4: IPNS
+
+Instead of using the DHT for storing the indexes, we could use IPNS where each peer would publish its list of signatures, and upon every signature, update what it's publishing.
```

Comment:
I think we could flesh this out a bit more, as I'm not sure how this works still.